map_annotator/nodes/map_annotator_node.py

```diff
@@ -16,11 +18,21 @@ def wait_for_time():
 class Server(object):
     def __init__(self):
         self.db = knowledge_representation.get_default_ltmc()
+        # subscriber and publisher
+        self.amcl_pose_sub = rospy.Subscriber("/amcl_pose", PoseWithCovarianceStamped, callback=self.publish_robot_pose)
+        self.robot_pose_pub = rospy.Publisher("map_annotator/robot_pose", RobotPose, queue_size=1)
+        # services
         self.get_maps_srv = rospy.Service("map_annotator/get_maps", GetMaps, self.get_all_maps)
         self.delete_map_srv = rospy.Service("map_annotator/delete_map", DeleteMap, self.delete_map)
         self.has_point_srv = rospy.Service("map_annotator/has_point", HasPoint, self.has_point)
         self.has_pose_srv = rospy.Service("map_annotator/has_pose", HasPose, self.has_pose)
         self.has_region_srv = rospy.Service("map_annotator/has_region", HasRegion, self.has_region)
+        rospy.sleep(0.5)
+
+
+    def publish_robot_pose(self, msg):
+        (roll, pitch, yaw) = euler_from_quaternion([msg.pose.pose.orientation.x, msg.pose.pose.orientation.y, msg.pose.pose.orientation.z, msg.pose.pose.orientation.w])
+        self.robot_pose_pub.publish(RobotPose(x=msg.pose.pose.position.x, y=msg.pose.pose.position.y, theta=yaw))
```

Comment:
The purpose of republishing the AMCL output is just to avoid having to do the quaternion to euler conversion in Javascript, right? It seems like some people import dependencies on the client side to handle this: https://github.com/OSUrobotics/contextual_interfaces/blob/c83c2d7936af7f10b60ed216586682ec33fac225/web/map.html#L77
But three.js seems like a lot to bring in for one function. Usually for such a well contained operation, I would try to extract the implementation out of the library.
It's possible you could use the roslibjs quaternion and apply the transformation to a unit vector to get theta: http://wiki.ros.org/roslibjs/Tutorials/MathFunctions

Reply:
Could you explain what do you mean by "apply the transformation to a unit vector to get theta"? I'm confused. Thank you!!

Reply:
Looking at the [ThreeVec API](http://robotwebtools.org/jsdoc/roslibjs/current/Vector3.html), I see that there's a `multiplyQuaternion` method. I think if you set up a vector pointing at theta=0 ( `0,1,0`?) then multiply it by a transformation, you can extract theta from the result by getting the angle between the original vector and the result with atan2.